### PR TITLE
Finalize Profile Skills page

### LINKED ISSUE #4859: Task - Finalize Applicant Profile Skills page.
# Description
This means creating a route to the page (with appropriate auth permissions), adding the React component to a twig template with the sidebar, checking that all links are correct, and ensuring it gets and deletes the Applicant's skills correctly.

# Images
[Skills Page wireframe](https://xd.adobe.com/view/c2c5a303-1dcd-4436-bec3-eea0a3ca3523-9796/)

# Acceptance criteria
- [x] Navigating to `profile/{applicantId}/skills` brings to to a page roughly matching the wireframe.
- [x] The above url is auth protected, so you can't visit if not the the correct user (admins excepted).
- [x] Navigating to `profile/skills` when logged in redirects you to the above url.
- [x] The breadcrumbs should read Home > Profile
- [x] The sidebar links work, and the My Skills entry is highlighted (and not a link).
- [x] The add skills button opens the Find and Explore Skills modal.
- [x] The applicant's skills are correctly displayed
- [x] After skills are added in the modal, they appear in the skills list.
- [ ] The Delete button deletes the skill on the server, and removes it from the UI.
- [ ] Passes Chrome's lighthouse tests for accessibility.

### PR DESCRIPTION
**Pull request by tristan-orourke**
Resolves #4859.

Unfortunately, a bunch of stuff got folded into this one PR.

### Primary changes
- Collects the Explore Skills modal and the Applicant Skills accordions on one page.
- Creates the route for the new page, with correct permissions, and sets up the correct API interactions.
- Some minor style changes to Skills List component.

### Related changes
- Client side Skill type now includes `skill_category_ids`, as is returned by the skills api endpoint. Changed some logic in List and FindSkillsModal components to use this to group skills.
- indexResourceHook now accepts a keyFn, which allows for mapping items by something other than id. This allows collecting Experiences (which include different types of experience, with overlapping ids) into one indexResourceHook.
- Created a new endpoint for all ExperienceSkills belonging to an applicant, so they can be retrieved with an indexResourceHook. (Previously they could only be retrieved along with the applicant experiences api endpoint.)
- Deleted some old Controller code for processing form post requests to update Degrees, Courses and WorkExperiences (deprecated).

### CODE REVIEW

File: resources/assets/js/components/ApplicantProfile/Skills/ProfileSkillsPage.tsx [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+  const submitNewSkills = async (newSkills: Skill[]): Promise<void> => {
+    await applicantSkillsResource.update({
+      skill_ids: [
+        ...applicantSkillsResource.value.skill_ids,
+        ...newSkills.map(getId),
+      ],
+    });
+  };
+
+  const removeSkill = async (skillId: number): Promise<void> => {
```
Comment (mnigh) [inline resources/assets/js/components/ApplicantProfile/Skills/ProfileSkillsPage.tsx:47]: This block of code is likely not the culprit, but I found an edge case where if there is only one skill attached to the applicant, it cannot be removed (422 Unprocessable Entity response is returned, see screenshot).

422 Unprocessable Entity response is returned

File: tests/Browser/helpers/constants.js [hunk 2]
```diff
@@ -21,7 +21,7 @@ export const MANAGER_FIRST_VISIT = `${HOMEPAGE}/manager/first-visit`;
 export const HR_FIRST_VISIT = `${HOMEPAGE}/hr/first-visit`;
 
 export const PROFILE_ABOUT = `${HOMEPAGE}/profile/about`;
-export const PROFILE_SKILLS = `${HOMEPAGE}/profile/skills`;
+export const PROFILE_SKILLS = `${HOMEPAGE}/profile/skills-old`;
```
Comment (mnigh) [inline tests/Browser/helpers/constants.js:24]: I'm not sure if this is meant to be part of this PR, but the Testcafe test for the profile has [not been updated](https://github.com/GCTC-NTGC/TalentCloud/blob/feature/profile-skills-finalized/tests/Browser/4-profile.testcafe.js#L18-L35) for the new skills page and so the test fails.

**Review by mnigh — CHANGES_REQUESTED** [review]

**Review by yonikid15 — COMMENTED** [review]
Wow it really come together nicely didn't it 👍 Overall it's looking really good
There are few things with the styling I would change though: 
- The titles a bit to big, I think setting the h2 heading property to h3 would look better. 
- The hover effect on the dialog trigger isn't working properly.
- It's hard to know that the skill accordions are accordions until you click on it. Maybe we could add a box-shadow effect with the H2 card property.

File: tests/Feature/Api/ExperienceSkillsControllerTest.php [hunk 3]
```diff
@@ -26,6 +26,64 @@ protected function makeExpSkillData($experienceId, $experienceType)
...
+            'experience_id' => $otherWork->id
+        ]);
+
+        $response = $this->actingAs($applicant->user)
+            ->get(route('api.v1.applicant.experience-skill.index', $applicant->id));
+        $response->assertOk();
+        $response->assertJsonCount(4);
+
+        // // Keys of each experience must be in alphabetical order for assertJsonFragment to match them.
+        // $expectedResponse = array_merge(
```
[the remark this answers is not in the document]
Comment (tristan-orourke) [reply tests/Feature/Api/ExperienceSkillsControllerTest.php:74]: Nope, this is left over from trying different things.

File: resources/assets/js/components/ApplicantProfile/Skills/List.tsx [hunk 4]
```diff
@@ -167,32 +176,60 @@ const List: React.FC<ListProps> = ({
...
-          </div>
-        ))}
+      {sortType === "group" && (
+        <>
+          {skillCategoriesGrouped.map((group) => (
+            <div key={group.category.id}>
+              <h5 data-h2-margin="b(top, 1) b(bottom, .5)">
+                {localizeFieldNonNull(locale, group.category, "name")}
+              </h5>
+              {group.skills
```
[the remark this answers is not in the document]
Comment (tristan-orourke) [reply resources/assets/js/components/ApplicantProfile/Skills/List.tsx:186]: Good call

File: tests/Browser/helpers/constants.js (hunk 2 above)
Comment (tristan-orourke) [reply tests/Browser/helpers/constants.js:24]: I'll create some issues to update testcafe tests for the new Application and Profile

**Comment by tristan-orourke** [comment]
> * The titles a bit to big, I think setting the h2 heading property to h3 would look better.

Done 👍 

> * The hover effect on the dialog trigger isn't working properly.

It looks okay to me. What behavior are you expecting that isn't there now?

> * It's hard to know that the skill accordions are accordions until you click on it. Maybe we could add a box-shadow effect with the H2 card property.

They did have card props, but they stopped working with an h2 update. 😆 I've replaced them with with the new `data-h2-shadow="b(medium"` prop, and spaced them out slightly more.

File: resources/assets/js/components/ApplicantProfile/Skills/ProfileSkillsPage.tsx (hunk 1 above)
Comment (tristan-orourke) [reply resources/assets/js/components/ApplicantProfile/Skills/ProfileSkillsPage.tsx:47]: Nice catch. Had to change the backend validation from `required` to `present`.

**Review by mnigh — COMMENTED** [review]
Nice work @tristan-orourke. This time around, I noticed a missing null state for skills when sort by alpha, so I've [added that](https://github.com/GCTC-NTGC/TalentCloud/pull/5192/commits/51d25af0e77ef69f26163cc217770e2d8013c613) as well. Otherwise, all of the issues I've mentioned have been addressed.

> The hover effect on the dialog trigger isn't working properly.

Like @tristan-orourke, I couldn't figure out what wasn't working properly. @yonikid15, could you elaborate or if the issue has resolved itself, let us know? After that, I can approve this to be merged.

**Comment by yonikid15** [comment]
> Nice work @tristan-orourke. This time around, I noticed a missing null state for skills when sort by alpha, so I've [added that](https://github.com/GCTC-NTGC/TalentCloud/pull/5192/commits/51d25af0e77ef69f26163cc217770e2d8013c613) as well. Otherwise, all of the issues I've mentioned have been addressed.
> 
> > The hover effect on the dialog trigger isn't working properly.
> 
> Like @tristan-orourke, I couldn't figure out what wasn't working properly. @yonikid15, could you elaborate or if the issue has resolved itself, let us know? After that, I can approve this to be merged.

Seems to be working now 👍

**Review by mnigh — APPROVED** [review]

**Pull request merged**